### PR TITLE
FIX: Update Numba Lecture to Address Deprecation of `@jit`

### LINKED ISSUE #294: Deprecation Warnings
@HumphreyYang would you mind clicking through the site and check for `Deprecation`

```bash
/tmp/ipykernel_4761/3271650557.py:1: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
```

there are a few places we need to fix up some warnings such as

https://python-programming.quantecon.org/numba.html#decorator-notation

### PR DESCRIPTION
**Pull request by HumphreyYang**
This PR resolves #294.

Future actions are held in https://github.com/QuantEcon/meta/issues/112

### CODE REVIEW

**Comment by HumphreyYang** [comment]
Hi @mmcky,

Could you please have a first pass on this?

Many thanks in advance.

File: lectures/numba.md [hunk 1]
```diff
@@ -246,29 +252,87 @@ This allows it to generate native machine code, without having to call the Pytho
...
+
+Note that a second signature with an `int64` array as the first argument is added into the signature of `bootstrap` function.
+
+The runtime is slower as if we ran the function for the first time.
+
+It suggests that Numba recompiles this function as the type changes.
+
+Overall, type inference helps Numba to achieve its performance, but it also limits what Numba supports as we have shown in the function example.
+
+In fact, this limitation means that Numba does not support everything in Python and scientific libraries we learned before.
```
Comment (mmcky) [inline lectures/numba.md]:
```suggestion
In fact, this limitation means that Numba does not support everything in Python and scientific libraries as we learned before.
```

**Comment by mmcky** [comment]
@HumphreyYang thanks for putting this together. This looks like a good change to me. 

I wonder if using `@jit(nopython=True)` would be useful to show instead of `@njit` as we will plan to swap back to using the simpler `@jit` directive once the deprecation has been made in a later release. 

Also can you please create and link the `switch back to @jit` issue. 

Thank you.

**Comment by HumphreyYang** [comment]
> @HumphreyYang thanks for putting this together. This looks like a good change to me.
> 
> I wonder if using `@jit(nopython=True)` would be useful to show instead of `@njit` as we will plan to swap back to using the simpler `@jit` directive once the deprecation has been made in a later release.
> 
> Also can you please create and link the `switch back to @jit` issue.
> 
> Thank you.

Hi @mmcky,

From the roadmap, I think Numba developers want to push people away from `@jit(nopython=False)`. If that is the case, I think `@njit` is more compact. Maybe we should ask @jstac for his opinion?

> Also can you please create and link the `switch back to @jit` issue.

It has been created in the `meta` repository [here](https://github.com/QuantEcon/meta/issues/112) as it might involve more than one lecture.

Many thanks in advance.

Best,
Humphrey

**Comment by jstac** [comment]
Thanks @HumphreyYang and @mmcky .  I'm indifferent between the two options, since we'll be changing both to @jit soon anyway.

**Comment by mmcky** [comment]
thanks @HumphreyYang -- that linked issue is useful so I have added it to the top level description. 

Do you know if `njit` is going to persist once the `@jit` decorator moves to default `nopython=True` mode?

**Comment by HumphreyYang** [comment]
> thanks @HumphreyYang -- that linked issue is useful so I have added it to the top level description.
> 
> Do you know if `njit` is going to persist once the `@jit` decorator moves to default `nopython=True` mode?

Many thanks @mmcky,

I checked their documentation, but I did not find anything about the deprecation of `njit`. There is no deprecation warning when using `njit` at the moment. Therefore, I think it is not going to be removed.

File: lectures/numba.md [hunk 2]
```diff
@@ -246,29 +252,87 @@ This allows it to generate native machine code, without having to call the Pytho
...
+
+Note that a second signature with an `int64` array as the first argument is added into the signature of `bootstrap` function.
+
+The runtime is slower as if we ran the function for the first time.
+
+It suggests that Numba recompiles this function as the type changes.
+
+Overall, type inference helps Numba to achieve its performance, but it also limits what Numba supports as we have shown in the function example.
+
+In fact, this limitation means that Numba does not support everything in Python and scientific libraries as we learned before.
```
Comment (mmcky) [inline lectures/numba.md]: @HumphreyYang I think we can delete this sentence as a repeat ... "In fact, .... learned before"

**Review by mmcky — COMMENTED** [review]
@HumphreyYang just one minor change. I am happy to keep `njit` for the sake of being concise.

**Review by mmcky — APPROVED** [review]
thanks @HumphreyYang 

@jstac there are some changes to content in this PR if you wanted to take a Quick Look.

File: lectures/numba.md [hunk 3]
```diff
@@ -246,29 +252,83 @@ This allows it to generate native machine code, without having to call the Pytho
...
+
+Let's try to change the data again and use an integer array as data
+
+```{code-cell} ipython3
+data = np.array([1, 2, 3, 4, 5], dtype=np.int64)
+%time bootstrap(data, mean, 100)
+bootstrap.signatures
+```
+
+Note that a second signature with is added.
```
Comment (jstac) [inline lectures/numba.md]: typo?

File: lectures/numba.md [hunk 4]
```diff
@@ -246,29 +252,83 @@ This allows it to generate native machine code, without having to call the Pytho
...
+
+It also takes longer to run suggesting that Numba recompiles this function as the type changes.
```
Comment (jstac) [inline lectures/numba.md]: "run" -> "run,"

**Comment by jstac** [comment]
Thanks @HumphreyYang , looking good.

@mmcky Could you please give this a careful proofread and then merge when ready?  I flagged a couple of small things above.

**Comment by HumphreyYang** [comment]
Many thanks @jstac and @mmcky ,

The typo is introduced when I modify the text according to @mmcky's feedback. I will be more careful when deleting content. I also found another typo in the original text : )

@mmcky, would you mind having another review and merge if it looks good?

Many thanks in advance.

**Pull request merged**